server-core/service/src/main/java/io/infectnet/server/service/impl/UserServiceImpl.java

```diff
@@ -1,22 +1,105 @@
 package io.infectnet.server.service.impl;
 
+import io.infectnet.server.persistence.Token;
+import io.infectnet.server.persistence.TokenStorage;
+import io.infectnet.server.persistence.User;
+import io.infectnet.server.persistence.UserStorage;
 import io.infectnet.server.service.UserDTO;
 import io.infectnet.server.service.UserService;
 
+import java.time.LocalDateTime;
+import java.util.Objects;
+import java.util.Optional;
+
 public class UserServiceImpl implements UserService {
 
+    private final UserStorage userStorage;
+
+    private final TokenStorage tokenStorage;
+
+    private final ConverterServiceImpl converterService;
+
+    public UserServiceImpl(){
+        converterService = null;
+        tokenStorage = null;
+        userStorage = null;
+    }
+
+    public UserServiceImpl(UserStorage userStorage, TokenStorage tokenStorage, ConverterServiceImpl converterService) {
+        this.userStorage = userStorage;
+        this.tokenStorage = tokenStorage;
+        this.converterService = converterService;
+    }
+
+    /**
+     * Registers the new user in. Checks if the token has not expired yet and the username is unique,
+     * and whether or not the password is valid according to the specified rules:
+     *      minimum length of 8 characters
+     *      must contain at least one number
+     *      must contain at least one uppercase letter.
+     * If authentication fails no new user will be stored, otherwise a new User is stored
+     * with the username, the email, the password and the date of the registration now produced.
+     *
+     * @param token the token from the current registration
+     * @param email the email address from the current registration
+     * @param username the username from the current registration
+     * @param password the password from the current registration
+     * @return the Optional containing the UserDTO made from the created user
+     */
     @Override
-    public UserDTO register(String token, String username, String password) {
-        return null;
+    public Optional<UserDTO> register(String token, String email, String username, String password) {
+        if(tokenIsValid(token) && emailIsUnique(email) && usernameIsUnique(username) && passwordIsValid(password)){
```

Comment:
Extract the contents of  this `if` into a separate method that checks for validity. Also modify the order of the clauses, so `tokenIsValid` comes first, followed by `passwordIsValid`.